### PR TITLE
7.0 add price propreties based

### PR DESCRIPTION
**Pull request by eLBati**
4 modules
## Easing properties input in sale order line

This modules simplifies the input of properties in the sale order line and
other places.

For instance, in the many2many field 'property_ids', it allows the user to
digit 'width 0.5' and the system will automatically create a property of group
'width' with value '0.5'

It also adds the model 'mrp.property.formula', to be used by computations based
on properties.
Used by modules like 'product_price_properties_based'
## Dynamic fields for sale order line properties

This module allows to dynamically draw properties groups on sale order lines.
That is, if you have a property group 'Length' and set it as 'Draw dynamically',
the module will automatically add the x_length field to the sale order line and
visualize it as char field in the form view.
When fill the 'Length' field, the module will update the properties field (property_ids)
according to it.
This allows to set the line's properties through normal fields, instead of creating
new properties or selecting existing properties.
## Product price properties based

This module allows to use python formaulas to compute the sale order line
price.

You can configure the 'Price formula' on the product form using python code.

Formula example:

```
area = float(properties['Width']) * float(properties['Length'])
result = area / 2.0
if 'Painting' in properties:
    result = result + 5
```

When changing properties on sale order line, the system will automatically
compute the line price unit.
## Sale line quantity properties based

_This module allows the calculation of the product quantity on the basis of a
formula that considers the properties specified by the user on the sale order
line and on the quantity (UoS)._

Example

Provided the sale of a given number of pieces (shelves), that may be
’x’ meter long and ’y’ meter large, the formula enables the calculation of the
total area sold expressed in square meters:
    10 pcs of (4 m x 0.5 m) shelves = 20 m² of wood

In order to have this function working, it is necessary to have the user
proceeding as follows:

Then s/he shall create properties such as ‘length 4’, ‘width 0.5’.
(Note: this can be more easily achieved by using the modules
'sale_properties_easy_creation' and/or 'sale_properties_dynamic_fields')

Properties must respond the following criteria:
    \* Name: ‘length 1’, ‘length 4’, ‘width 0.5’
    \* Property Group : either ‘length’ or ‘width’
    \* Value : the corresponding quantity (1, 4, 0.5...)

Property 'length 4'
    \* Name: ‘length 4’
    \* Property Group : ‘length’
    \* Value : 4

Property 'width 0.5'
    \* Name: ‘width 0.5’
    \* Property Group : ‘width’
    \* Value : 0.5

After this, the formula 'surface' must be created and associated
to the product:

```
result = float(properties['length']) * float(properties['width']) * qty_uos
```

Upon the registering of the order, the user will apply in the properties field
the desired properties (in this example the ‘lenght 4’ and ‘width 2’), the
needed formula (in this example 'surface') and last the quantity (UoS).

### CODE REVIEW

**Comment by rvalyi** [comment]
Hello,

this looks good features overall.

Still, I see you are replacing the original product_id_change of 3 fields with your new product_id_change_with_properties on_change event.

Well, this is brittle and will be incompatible with other modules trying the same thing.
So I currently vote against the merge :-1: 

What about depending on web_context_tunnel instead, leave the original product_id_change method and inject your extra properties field with a new specially context key.

See an example here for Brazilian localization:
https://github.com/openerpbrasil/l10n_br_core/blob/develop/l10n_br_sale/sale_view.xml#L36
Also read https://github.com/OCA/server-tools/blob/7.0/web_context_tunnel/__openerp__.py

What do you think?

**Comment by eLBati** [comment]
@rvalyi thanks for the valuable suggestion. It's a very efficient solution.

I just had to modify the test, because simulating the loading of the view through yaml

```
!record {model: sale.order, id: sale_order_test1, view: sale.view_order_form}:
```

did not trigger the tunnel: I got an empty context

**Comment by rvalyi** [comment]
looks good, having to hack the tests is expected indeed. I'm doing a few tests and hopefully will approve. Sorry for the delay and thank you for the awesome work!

**Comment by rvalyi** [comment]
Hello @eLBati 
so I took some time to test all the 3 modules and I think they are great and greatly modularized into atomic features. I also worked on product configurator some times ago, these modules will definitely be a solid foundation for this.

So congratulations and :+1:

**Comment by eLBati** [comment]
@rvalyi many thanks for your tests and review.

I just added another similar module: `sale_line_quantity_properties_based`
It works like `sale_line_price_properties_based` but it computes the `Quantity` field.
See the updated PR description

File: sale_line_price_properties_based/sale.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+
+from openerp.osv import orm
+from openerp.tools.translate import _
+
+
+class SaleOrderLine(orm.Model):
+    _inherit = 'sale.order.line'
```
Comment (bwrsandman) [inline sale_line_price_properties_based/sale.py]: Please name the file after the model/class: `sale_order_line.py`

File: sale_line_price_properties_based/sale_demo.xml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="utf-8"?>
+<openerp>
+    <data>
+        <record model="product.product" id="product.product_product_7">
```
Comment (bwrsandman) [inline sale_line_price_properties_based/sale_demo.xml]: Please name the demo file after the model: `product_demo.xml` or `product_product_demo.xml`

File: sale_line_price_properties_based/sale_view.xml [hunk 3]
```diff
@@ -0,0 +1,24 @@
+<?xml version="1.0" encoding="utf-8"?>
+<openerp>
+    <data>
+
+        <record id="view_order_form" model="ir.ui.view">
+            <field name="name">sale.order.form</field>
+            <field name="model">sale.order</field>
```
Comment (bwrsandman) [inline sale_line_price_properties_based/sale_view.xml]: Please name the view after the model: `sale_order_view.xml`

File: sale_line_quantity_properties_based/sale.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+from openerp.osv import orm
+from openerp.tools.translate import _
+import logging
+
+_logger = logging.getLogger(__name__)
+
+
+class SaleOrderLine(orm.Model):
+    _inherit = 'sale.order.line'
```
Comment (bwrsandman) [inline sale_line_quantity_properties_based/sale.py]: Please name the file after the model/class: `sale_order_line.py`

File: sale_line_quantity_properties_based/sale_view.xml [hunk 5]
```diff
@@ -0,0 +1,37 @@
+<?xml version="1.0" encoding="utf-8"?>
+<openerp>
+    <data>
+
+        <record id="view_order_form_inherit" model="ir.ui.view">
+            <field name="name">sale.order.form.sale.stock</field>
+            <field name="model">sale.order</field>
```
Comment (bwrsandman) [inline sale_line_quantity_properties_based/sale_view.xml]: Please name the view after the model: `sale_order_view.xml`

File: sale_properties_dynamic_fields/mrp_property.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+
+from openerp.osv import orm, fields
+from openerp.tools.translate import _
+
+
+class MrpPropertyGroup(orm.Model):
+    _inherit = 'mrp.property.group'
```
Comment (bwrsandman) [inline sale_properties_dynamic_fields/mrp_property.py]: Please name file after the model/class: `mrp_property_group.py`

File: sale_properties_dynamic_fields/mrp_property_view.xml [hunk 7]
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0" encoding="utf-8"?>
+<openerp>
+    <data>
+        
+        <record id="mrp_property_group_form_view_draw" model="ir.ui.view">
+            <field name="name">mrp_property_group_form_view_draw</field>
+            <field name="model">mrp.property.group</field>
```
Comment (bwrsandman) [inline sale_properties_dynamic_fields/mrp_property_view.xml]: Please name the view after the model: `mrp_property_group_view.xml`

File: sale_properties_dynamic_fields/sale.py [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+#
+##############################################################################
+
+from openerp.osv import orm
+from lxml import etree
+from openerp.tools.translate import _
+
+
+class SaleOrder(orm.Model):
+    _inherit = 'sale.order'
```
Comment (bwrsandman) [inline sale_properties_dynamic_fields/sale.py]: Please put one model/class per python script: `sale_order.py`

File: sale_properties_dynamic_fields/sale.py [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+                        "//field[@name='property_ids']")[0]
+                    prop_m2m_field.addprevious(group_field)
+                    res['fields']['order_line']['views']['form'][
+                        'arch'
+                        ] = etree.tostring(eview, pretty_print=True)
+        return res
+
+
+class SaleOrderLine(orm.Model):
+    _inherit = 'sale.order.line'
```
Comment (bwrsandman) [inline sale_properties_dynamic_fields/sale.py]: Please put one model/class per python script: `sale_order_line.py`

File: sale_properties_dynamic_fields/test/properties.yml [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+# KeyError: "Object '<openerp.osv.orm.sale.order.line object at 0xb2b67a4c>' does not contain field 'x_width'"
+#-
+#  I create a sale order
+#-
+#  !record {model: sale.order, id: sale_order_test1, view: sale.view_order_form}:
+#    partner_id: base.res_partner_2
+#    order_line: 
+#      - product_id: product.product_product_7
+#        product_uom_qty: 8
+#        x_width: 7
```
Comment (bwrsandman) [inline sale_properties_dynamic_fields/test/properties.yml:69]: This is all commented out, is there a point to commiting this or can it be taked out?

File: sale_properties_easy_creation/mrp_property.py [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+                    'group_id': group_ids[0],
+                    'value': splitted_name[1]
+                    }, context=context)
+                return self.name_get(cr, uid, [rec_id], context)[0]
+        return super(mrp_property, self).name_create(
+            cr, uid, name, context=context)
+
+
+class MrpPropertyFormula(orm.Model):
+    _name = 'mrp.property.formula'
```
Comment (bwrsandman) [inline sale_properties_easy_creation/mrp_property.py]: Please put one model/class per python script: `mrp_property_formula.py`

File: sale_properties_easy_creation/mrp_property_view.xml [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+                <field name="composition" position="after">
+                    <field name="value"></field>
+                </field>
+            </field>
+        </record>
+        
+        <!-- formula -->
+        <record id="mrp_property_formula_tree_view" model="ir.ui.view">
+            <field name="name">mrp.property.formula.tree</field>
+            <field name="model">mrp.property.formula</field>
```
Comment (bwrsandman) [inline sale_properties_easy_creation/mrp_property_view.xml]: Please put one model per view file: `mrp_property_formula_view.xml`

File: sale_properties_easy_creation/sale_demo.xml [hunk 13]
```diff
@@ -0,0 +1,41 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (bwrsandman) [inline sale_properties_easy_creation/sale_demo.xml]: Please put one model per demo file and name the file after the model.

There is nothing to do with the sale model here.

File: sale_properties_dynamic_fields/test/properties.yml (hunk 10 above)
Comment (eLBati) [inline sale_properties_dynamic_fields/test/properties.yml:69]: I kept it as reference about how the test should be done if yaml_import.py worked properly.
That is: if and when yaml_import.py will be able to handle it, the test will can be activated

Comment (bwrsandman) [inline sale_properties_dynamic_fields/test/properties.yml:69]: I doubt another dev would uncomment it even if adding tests once yaml_import.py is fixed.

File: sale_properties_easy_creation/sale_demo.xml (hunk 13 above)
Comment (bwrsandman) [inline sale_properties_easy_creation/sale_demo.xml]: I think you forgot to remove this file

**Comment by bwrsandman** [comment]
Code review: LGTM :+1: 

There is a file left over that is no longer user: sale_properties_easy_creation/sale_demo.xml
I think you might have forgotten to remove it

**Comment by eLBati** [comment]
Also removed sale_properties_easy_creation/sale_demo.xml
Thanks

File: sale_line_price_properties_based/product.py [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+                 " - cr\n"
+                 " - uid\n"
+                 " - ptype\n"
+                 " - product_id\n"
+                 " - properties (dictionary of properties)\n"
+                 "You have to put the result in the 'result' variable"),
+        }
+
+    def price_get(self, cr, uid, ids, ptype='list_price', context=None):
+        if 'properties' in context:
```
Comment (guewen) [inline sale_line_price_properties_based/product.py:49]: `context` may be None, needs to be checked before accessed

File: sale_properties_dynamic_fields/mrp_property_group.py [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+
+from openerp.osv import orm, fields
+from openerp.tools.translate import _
+
+
+class MrpPropertyGroup(orm.Model):
+    _inherit = 'mrp.property.group'
+    _columns = {
+        'draw_dynamically': fields.boolean(
+            'Draw dynamically', help="In nsale order line, draw this property "
```
Comment (guewen) [inline sale_properties_dynamic_fields/mrp_property_group.py]: s/nsale/sale/

**Comment by guewen** [comment]
Added a few line comments. Globally, I find the indentation really weird, but I won't argue on that.

**Comment by eLBati** [comment]
Thanks @guewen, fixed your remarks.

About indentation, would you indicate a weird example and how would you modify it?
Thanks

File: sale_properties_easy_creation/__openerp__.py [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+This modules simplifies the input of properties in the sale order line and
+other places.
+
+For instance, in the many2many field 'property_ids', it allows the user to
+digit 'width 0.5' and the system will automatically create a property of group
+'width' with value '0.5'
+
+It also adds the model 'mrp.property.formula', to be used by computations based
+on properties.
+Used by modules like 'product_price_properties_based' and
```
Comment (tafaRU) [inline sale_properties_easy_creation/__openerp__.py]: This module seems do not exist, maybe it's a typo ;-)

File: sale_line_price_properties_based/i18n/it.po [hunk 17]
```diff
@@ -0,0 +1,63 @@
+# Translation of OpenERP Server.
+# This file contains the translation of the following modules:
+#	* product_price_properties_based
```
Comment (tafaRU) [inline sale_line_price_properties_based/i18n/it.po]: It should be referred to 'sale_line_price_properties_based' instead 'product_price_properties_based'
Same error in the following lines.

Comment (eLBati) [inline sale_line_price_properties_based/i18n/it.po]: @tafaRU fixed references to old module name.
Thanks

**Comment by eLBati** [comment]
OCB tests are red because of https://github.com/OCA/OCB/issues/37

**Comment by eLBati** [comment]
Hello @rvalyi can you take a look at the 2 other modules I added after your review?

**Comment by rvalyi** [comment]
@eLBati will try to find time. Not possible today nor tomorrow sadly.

**Comment by rvalyi** [comment]
:+1: LGTM

sorry for the delay. Hope we can proceed with the merge then

**Pull request merged**